protocols/gossipsub/src/handler.rs

```diff
@@ -564,17 +516,58 @@ impl ConnectionHandler for Handler {
             Self::OutboundOpenInfo,
         >,
     ) {
+        if event.is_outbound() {
+            self.outbound_substream_establishing = false;
+        }
+
+        if event.is_inbound() && self.inbound_substreams_created == MAX_SUBSTREAM_CREATION {
+            // Too many inbound substreams have been created, disable the handler.
+            self.keep_alive = KeepAlive::No;
+            log::info!("The maximum number of inbound substreams created has been exceeded.");
```

Comment:
Maybe this should be an error or a crit. In principle this should never happen and we don't want the connection to just end silently

Reply:
The thinking was that `info!` is typically used for a state change (we are disabling the handler). I don't want to wake an ops person at 3am because we are emitting `warn` or `error` here :)

Happy to be convinced otherwise.

Reply:
Yeah, I'm not all that concerned. I;d be curious to know when to use a warn then. I'd typically use a warn if something has not behaved as it should and is worth notifying a user. 
In this case, we pretty much should never see this log, unless something is pretty broken. I find it handy to grep warn/error etc to find things that are broken. This fits into that classification imo. 

But if other parts of libp2p are not doing it, happy to leave as an info.

Reply:
Why would we never see this log? We would see this if we connect to a node that doesn't support the gossipsub protocol right? Nothing is inherently broken if that happens so I am not sure a `warn` is appropriate.

Reply:
Agreed that this should rather be a `warn` instead of a `info`.

> Why would we never see this log?

By default `env-logger` does not log on `info` level.

> In this case, we pretty much should never see this log, unless something is pretty broken. I find it handy to grep warn/error etc to find things that are broken. This fits into that classification imo.

Agreed.

> We would see this if we connect to a node that doesn't support the gossipsub protocol right?

As far as I can tell a `UpgradeError::Select(NegotiationError::Failed)` would lead to `self.protocol_unsupported = true` and thus a direct disabling (i.e. `self.keep_alive = KeepAlive::No`) in the next `ConnectionHandler::poll` invocation. Thus we would never request another outbound stream and thus the log line above is never printed. Am I missing something @thomaseizinger?

Just to avoid confusion, I am in favor of the above behavior. I don't think there is value in directly requesting another stream in case the remote signaled that it does not support the protocol on the previous stream. Objections?

Reply:
Makes sense!